observability/src/main/java/io/grpc/observability/logging/GcpLogSink.java

```diff
@@ -0,0 +1,155 @@
+  // it as an argument
+  public static synchronized GcpLogSink init(String projectId) {
+    if (gcpLogSinkInstance != null) {
+      throw new IllegalStateException("GcpLogSink already initialized");
+    }
+    gcpLogSinkInstance = new GcpLogSink(projectId);
+    return gcpLogSinkInstance;
+  }
+
+  /**
+   * Retrieves a single instance of GcpLogSink.
+   */
+  public static GcpLogSink getInstance() {
+    if (gcpLogSinkInstance == null) {
+      throw new IllegalStateException("GcpLogSink not initialized");
+    }
+    return gcpLogSinkInstance;
+  }
+
+  @Override
+  public synchronized void write(GrpcLogRecord logProto) {
+    if (gcpLoggingClient == null) {
+      logger.log(Level.SEVERE, "Attempt to write after GcpLogSink is closed.");
+      return;
+    }
+    if (SERVICE_TO_EXCLUDE.equals(logProto.getServiceName())) {
+      return;
+    }
+    try {
+      Map<String, Object> mapPayload = protoToMapConverter(logProto);
+      Severity logEntrySeverity = getCloudLoggingLevel(logProto.getLogLevel());
+      // TODO(vindhyan): make sure all (int, long) values are not displayed as double
+      LogEntry grpcLogEntry =
+          LogEntry.newBuilder(JsonPayload.of(mapPayload))
+              .setSeverity(logEntrySeverity)
+              .setLogName(DEFAULT_LOG_NAME)
+              .setResource(MonitoredResource.newBuilder("global").build())
+              .build();
+      gcpLoggingClient.write(Collections.singleton(grpcLogEntry));
+    } catch (Exception e) {
+      logger.log(Level.SEVERE, "Caught exception while writing to cloud logging", e);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Map<String, Object> protoToMapConverter(GrpcLogRecord logProto)
+      throws InvalidProtocolBufferException, IOException {
+    JsonFormat.Printer printer = JsonFormat.printer().preservingProtoFieldNames();
+    String recordJson = printer.print(logProto);
+    return (Map<String, Object>) JsonParser.parse(recordJson);
+  }
+
+  private Severity getCloudLoggingLevel(GrpcLogRecord.LogLevel recordLevel) {
+    switch (recordLevel.getNumber()) {
+      case 1: // GrpcLogRecord.LogLevel.LOG_LEVEL_TRACE
+      case 2: // GrpcLogRecord.LogLevel.LOG_LEVEL_DEBUG
+        return Severity.DEBUG;
+      case 3: // GrpcLogRecord.LogLevel.LOG_LEVEL_INFO
+        return Severity.INFO;
+      case 4: // GrpcLogRecord.LogLevel.LOG_LEVEL_WARN
+        return Severity.WARNING;
+      case 5: // GrpcLogRecord.LogLevel.LOG_LEVEL_ERROR
+        return Severity.ERROR;
+      case 6: // GrpcLogRecord.LogLevel.LOG_LEVEL_CRITICAL
+        return Severity.CRITICAL;
+      default:
+        return Severity.DEFAULT;
+    }
+  }
+
+  /**
+   * Closes Cloud Logging Client.
+   */
+  public synchronized void close() {
+    if (gcpLoggingClient == null) {
+      return;
+    }
+    try {
+      gcpLoggingClient.close();
+    } catch (Exception e) {
+      logger.log(Level.SEVERE, "Caught exception while closing", e);
+    } finally {
+      gcpLogSinkInstance = null;
```

Comment:
In non-static method `close()` you are setting the static member `gcpLogSinkInstance` to null - which is not completely wrong but I think we can do better. I notice that `GcpLogSink getInstance()` is called only to call close on it. So we can rename `getInstance()` to something like `finish()` (or deInit()) and do:

```
  public static synchronized GcpLogSink finish() {
    if (gcpLogSinkInstance == null) {
      throw new IllegalStateException("GcpLogSink not initialized");
    }
    gcpLogSinkInstance.close();
    gcpLogSinkInstance = null;
  }
```

and change `close()` to be private and remove the line setting `gcpLogSinkInstance` to null there.

Reply:
For a singleton class, I thought it would not be right to set static instance to null, so not setting it to null for now.